connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncStore.java

```diff
@@ -25,17 +25,37 @@
 import org.apache.kafka.common.utils.Utils;
 import org.apache.kafka.connect.util.KafkaBasedLog;
 import org.apache.kafka.connect.util.TopicAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Optional;
 import java.util.OptionalLong;
 import java.util.concurrent.ConcurrentHashMap;
 
-/** Used internally by MirrorMaker. Stores offset syncs and performs offset translation. */
+/**
+ * Used internally by MirrorMaker. Stores offset syncs and performs offset translation.
+ * <p>A limited number of offset syncs can be stored per TopicPartition, in a way which provides better translation
+ * later in the topic, closer to the live end of the topic.
+ * This maintains the following invariants for each topic-partition in the in-memory sync storage:
+ * <ul>
+ *     <li>syncs[0] is the latest offset sync from the syncs topic</li>
+ *     <li>For each i,j, i <= j: syncs[j].upstream <= syncs[i].upstream < syncs[j].upstream + 2^j</li>
+ * </ul>
+ * <p>Offset translation uses the syncs[i] which most closely precedes the upstream consumer group's current offset.
+ * For a fixed in-memory state, translation of variable upstream offsets will be monotonic.
+ * For variable in-memory state, translation of a fixed upstream offset will not be monotonic.
+ * <p>Translation will be unavailable for all topic-partitions before an initial read-to-end of the offset syncs topic
+ * is complete. Translation will be unavailable after that if no syncs are present for a topic-partition, or if relevant
+ * offset syncs for the topic were eligible for compaction at the time of the initial read-to-end.
```

Comment:
> it's also possible that replication began from a later point than the consumer group is reading from, which is also a possible reason that translation may be unavailable. Oh, and I guess there's the case where replication began, but was unable to produce offset syncs for whatever reason

Thanks for pointing that out, I missed those conditions.

> does it cover the case where syncs for earlier offsets have been wiped due to compaction? If so, we may want to clarify that offset syncs were "lost due to compaction" instead of "eligible for compaction".

It does cover the cases where syncs are lost to compaction, but eligible for compaction includes even more records. This is alluding to the `!readToEnd` condition that you pointed out in another comment.

Reply:
Does the part of the Javadoc on the availability of translation still need to be updated?